File: CivitasJava/civitas/SorpresaSalirCarcel.java
package civitas;

import java.util.ArrayList;

public class SorpresaSalirCarcel extends Sorpresa {
    private MazoSorpresas mazo;
    
    SorpresaSalirCarcel(MazoSorpresas mazo, String tex){
        super(tex);
        this.mazo = mazo;
    }
    
    @Override
    void aplicarAJugador (int actual, ArrayList<Jugador> todos) throws Exception{
        if ( jugadorCorrecto(actual, todos) ){
            informe(actual, todos);
            
            boolean tieneSalvoconducto = false;
            for (int i = 0; i < todos.size(); i++)
                if (todos.get(i).tieneSalvoConducto() )
                    tieneSalvoconducto = true;
            
            if (!tieneSalvoconducto){
                todos.get(actual).obtenerSalvoconducto(this);
                salirDelMazo();
            }
        }
    }
    
    void salirDelMazo(){
        mazo.inhabilitarCartaEspecial(this);
    }
    
    void usada(){
        mazo.habilitarCartaEspecial(this);
    }
}

File: CivitasRuby/sorpresa_salir_carcel.rb
require_relative 'sorpresa'
require_relative 'mazo_sorpresas'

module Civitas
class SorpresaSalirCarcel
  
  def initialize(mazo, texto)
    super()
    @mazo = mazo
    @texto = texto
  end
  
   def aplicar_a_jugador(actual, todos)
    if jugador_correcto(actual, todos)
      informe(actual, todos)
      
      tiene_salvoconducto = false
      for i in 0..3
        if (todos[i].tiene_salvoconducto)
          tiene_salvoconducto = true
        end
      end
      
      if (!tiene_salvoconducto)
        todos[actual].obtener_salvoconducto
        salir_del_mazo
      end
    end
  end
  
  def usada
    if @tipo == TipoSorpresa::SALIRCARCEL
      @mazo.habilitar_carta_especial(self)
    end
  end
  
  def  salir_del_mazo
    if @tipo == TipoSorpresa::SALIRCARCEL
      @mazo.inhabilitar_carta_especial(self)
    end
  end
  
  public_class_method :new
end
end
